File: lc_cpp/lcCpp/dp/maxSubarraySumCircular.cpp
//
// Created by 24276 on 2025/8/12.
//

#include <vector>
using namespace std;
class max_subarray_sum_circular {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int total = 0, mxEn = 0, meEn = 0, mxSum = nums[0], miSum = nums[0];
        for (int v : nums ) {
            total += v;
            mxEn = max(v, v + mxEn), mxSum = max(mxSum, mxEn),
            meEn = min(v, v + meEn), miSum = min(miSum, meEn);

        }
        if (mxSum < 0) return mxSum;
        return max(mxSum, total - miSum);
    }
};

int main() {

    return 0;
}

File: lc_py/lcPy/dp/maxSubarraySumCircular.py
from typing import List


class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        total, mx_sum, mi_sum, mx_en, mi_en = 0, nums[0], nums[0], 0, 0

        for v in nums:
            total += v
            mx_en = max(v, v + mx_en)
            mx_sum = max(mx_sum, mx_en)
            mi_en = min(v, v + mi_en)
            mi_sum = min(mi_sum, mi_en)
        return mx_sum if mx_sum < 0 else max(mx_sum, total - mi_sum)